src/wlapplication_options.cc

```diff
@@ -826,6 +820,38 @@ KeyboardShortcut shortcut_from_string(const std::string& name) {
 	throw wexception("Shortcut '%s' does not exist", name.c_str());
 }
 
+std::string keymod_string_for(const uint16_t modstate, const bool rt_escape) {
+	i18n::Textdomain textdomain("widelands");
+	std::vector<std::string> mods;
+	if (modstate & KMOD_SHIFT) {
+		mods.push_back(pgettext("hotkey", "Shift"));
+	}
+	if (modstate & KMOD_ALT) {
+		mods.push_back(pgettext("hotkey", "Alt"));
+	}
+	if (modstate & KMOD_GUI) {
+#ifdef __APPLE__
+		mods.push_back(pgettext("hotkey", "Cmd"));
+#else
+		mods.push_back(pgettext("hotkey", "GUI"));
+#endif
+	}
+	if (modstate & KMOD_CTRL) {
+		mods.push_back(pgettext("hotkey", "Ctrl"));
+	}
+
+	std::string result;
+
+	// Return value will have a trailing "+" if any modifier is matched,
+	// because all current uses need it anyway, and extra checks can
+	// be avoided both here and in the users this way
+	for (const std::string& m : mods) {
+		result = (boost::format(_("%1$s+%2$s")) % m % result).str();
```

Comment:
In RTL languages this generates a string like `+GUI+Ctrl`. For a keyboard shortcut this is fine since that will result in `X+GUI+Ctrl`, but in the mousewheel UI it will look like
> Zoom Map: &nbsp;&nbsp;&nbsp;&nbsp; +Gui+Ctrl &nbsp;&nbsp;&nbsp;&nbsp; Any Scroll

Reply:
Uh-oh... didn't think of that. Ideally the whole layout of the hbox should be flipped, but is there a way to tell from the locale whether it is LTR or RTL? Other than that, ":paM mooZ   iuG+lrtC   llorcS ..." isn't much better... (BTW spinbox should also be flipped for RTL, but I don't think it is - I checked it, and it turns out that all RTL languages have very low translation coverage, so they mostly show the original English strings in Options...)

Is there a way to handle this situation?

Reply:
Not really… ideally the horizontal layout should be mirrored for RTL languages, but for now support for this is veeery patchy. It can be handled like this:
https://github.com/widelands/widelands/blob/eccffa9b227196cec5471f3a14ebfc854f3da970/src/editor/ui_menus/main_menu_random_map.cc#L762-L768
We also have some "convenience" functions for use with Align-layouted boxes:
https://github.com/widelands/widelands/blob/eccffa9b227196cec5471f3a14ebfc854f3da970/src/graphic/align.cc#L47-L52
The inofficial policy is to implement it in this way for additions to the UI if the reviewer remembers to ask for it ;)

Reply:
Do I get it right that `UI::g_fh` depends on the global locale setting, so it will be wrong if a given string has no translation? But with low translation coverage this will be a mess if I reorder everything? I guess here I could reorder based on the direction of the title, so if locale is RTL, I need to check whether the title has a translation. What's the best way to do it other than comparing the raw string to the translated one?

And `mirror_alignment()` is intended to be used in `add()`, so I still need to conditionally reorder the widgets in the box?

Reply:
If we try to account for translation coverage in RTL support the layouting code will quickly become unmaintainable. It's more viable (and efficient) to just assume that all languages are fully translated and only query the global fontset.

Usually, you *only* need to conditionally reorder the order in which widgets are added to a `Box`. The `mirror_alignment` function would be needed for widgets that are added with `Box::Resizing::kAlign` to a `Vertical` box – this would then look like
```c++
vbox.add(some_panel, UI::Box::Resizing::kAlign, mirror_alignment(UI::Align::kLeft));
```
We don't yet have any code examples that use this though. 
`Resizing::kAlign` is usually **not** what you want because it does not resize the widget to fill available space; `kFullSize` or `kExpandBoth` is often the better choice. You could try using these for the options menu boxes – at 1024×768 the tab looks nice and compact, but at 1920×1080 there's masses of empty space which could be filled by horizontally expanding the dropdowns.
Unfortunately `kAlign`+`kLeft` is the default if only one argument is passed to `add`.

Reply:
I found that multilinetextarea uses `has_rtl_character()` from `graphic/text/bidi.h` as condition in `mirror_alignment()`. I think reordering `add()`-s based on that wouldn't be too messy. (or more messy than reordering based on global setting anyway)

Hmmm, try it with spinbox in another PR? At least that can be tested with Options in (I think) Arabic...
![shot0000](https://user-images.githubusercontent.com/80712255/135749344-cb6be825-3e03-469e-af89-5943fd680a74.png)
Though spinbox is quite messy as it is with the conditional big step buttons...

Reply:
`has_rtl_character` has very bad performance, ideally it should be used only for text rendering internals and other code should assume that the fontset provides all the information we need.

Reply:
I tried `UI::g_fh->fontset()->is_rtl()`, but it only works for hindi. Arabic, farsi and hebrew get false.

Reply:
Hm, that's weird. I tested my `techinfo` branch with Hebrew and it inverted the layout as expected.
The text direction is defined in `data/i18n/fonts.lua` which declares RTL for Arabic, Devanagari, and Hebrew font sets, used – according to `data/i18n/locales.lua` by Al-ʿArabiyyah, Farsi, Hindi, and Ivrit.

Reply:
I was able to find the languages because they are shown right aligned in the language selection dropdown.
Hmmm, may this be because I'm using Debian with fonts provided by its packages?

Reply:
Anyway, it's an unrelated issue... Done.